src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,889 @@
+    A = randn(n, n)
+    B = randn(n, n)
+    C = randn(n, n)
+    B = (1 / 2) .* (B .+ B')
+    C = (1 / 2) .* (C .+ C')
+    Ω = [A B; C -A']
+    return final_norm * Ω / norm(Ω, 2)
+end
+
+@doc raw"""
+    retract(::Symplectic, p, X, ::CayleyRetraction)
+    retract!(::Symplectic, q, p, X, ::CayleyRetraction)
+
+Compute the Cayley retraction on ``p ∈ \operatorname{Sp}(2n, ℝ)`` in
+the direction of tangent vector ``X ∈ T_p\operatorname{Sp}(2n, ℝ)``,
+as defined in by Birtea et al in proposition 2 [^birtea2020optimization].
+
+Using the symplectic inverse of a matrix ``A \in ℝ^{2n \times 2n}``,
+``
+A^{+} := Q_{2n}^T A^T Q_{2n}
+``
+where
+````math
+Q_{2n} =
+\begin{bmatrix}
+0_n & I_n \\
+ -I_n & 0_n
+\end{bmatrix},
+````
+the retraction
+``\mathcal{R}\colon T\operatorname{Sp}(2n) \rightarrow \operatorname{Sp}(2n)``
+is defined pointwise as
+````math
+\begin{align*}
+\mathcal{R}_p(X) &= p \operatorname{cay}\left(\frac{1}{2}p^{+}X\right), \\
+                 &= p \operatorname{exp}_{1/1}(p^{+}X), \\
+                 &= p (2I - p^{+}X)^{-1}(2I + p^{+}X).
+\end{align*}
+````
+Here
+``\operatorname{exp}_{1/1}(z) = (2 - z)^{-1}(2 + z)``
+denotes the Padé (1, 1) approximation to ``\operatorname{exp}(z)``.
+
+[^birtea2020optimization]:
+    > Birtea, Petre and Ca{\c{s}}u, Ioan and Com{\u{a}}nescu, Dan:
+    > Optimization on the real symplectic group,
+    > Monatshefte f{\"u}r Mathematik, Springer
+    > doi [10.1007/s00605-020-01369-9](https://doi.org/10.1007/s00605-020-01369-9)
+"""
+function retract!(M::Symplectic, q, p, X, ::CayleyRetraction)
+    p_star_X = symplectic_inverse_times(M, p, X)
+
+    divisor = lu(2 * I - p_star_X)
+    q .= p * (divisor \ add_scaled_I!(p_star_X, 2.0))
+    return q
+end
+
+Base.show(io::IO, ::Symplectic{n,𝔽}) where {n,𝔽} = print(io, "Symplectic{$(2n), $(𝔽)}()")
+
+function symplectic_inverse_times(M::Symplectic{n}, p, q) where {n}
+    A = similar(p)
+    return symplectic_inverse_times!(M, A, p, q)
+end
+
+function symplectic_inverse_times!(::Symplectic{n}, A, p, q) where {n}
+    # we write p = [p1 p2; p3 p4] (and q, too), then
+    p1 = @view(p[1:n, 1:n])
+    p2 = @view(p[1:n, (n + 1):(2n)])
+    p3 = @view(p[(n + 1):(2n), 1:n])
+    p4 = @view(p[(n + 1):(2n), (n + 1):(2n)])
+    q1 = @view(q[1:n, 1:n])
+    q2 = @view(q[1:n, (n + 1):(2n)])
+    q3 = @view(q[(n + 1):(2n), 1:n])
+    q4 = @view(q[(n + 1):(2n), (n + 1):(2n)])
+    A1 = @view(A[1:n, 1:n])
+    A2 = @view(A[1:n, (n + 1):(2n)])
+    A3 = @view(A[(n + 1):(2n), 1:n])
+    A4 = @view(A[(n + 1):(2n), (n + 1):(2n)])
+    mul!(A1, p4', q1) # A1 = p4'q1
+    mul!(A1, p2', q3, -1, 1) # A1 -= p2'p3
+    mul!(A2, p4', q2) # A2 = p4'q2
+    mul!(A2, p2', q4, -1, 1) #A2 -= p2'q4
+    mul!(A3, p1', q3) #A3 = p1'q3
+    mul!(A3, p3', q1, -1, 1) # A3 -= p3'q1
+    mul!(A4, p1', q4) # A4 = p1'q4
+    mul!(A4, p3', q2, -1, 1) #A4 -= p3'q2
+    return A
+end
+
+ndims(Q::SymplecticMatrix) = 2
+copy(Q::SymplecticMatrix) = SymplecticMatrix(Q.λ)
```

Comment:
```suggestion
copy(Q::SymplecticMatrix) = SymplecticMatrix(copy(Q.λ))
```
Most likely makes no difference in nearly any situation but this seems more appropriate.

Reply:
:+1: